File: com.systemincloud.examples.tasks.javatask/src/test/java/tasks/extsrc/ExtSrcAtomicWith1SIn.java
package tasks.extsrc;

import com.systemincloud.modeler.tasks.javatask.api.InputPort;
import com.systemincloud.modeler.tasks.javatask.api.JavaTask;
import com.systemincloud.modeler.tasks.javatask.api.OutputPort;
import com.systemincloud.modeler.tasks.javatask.api.annotations.InputPortInfo;
import com.systemincloud.modeler.tasks.javatask.api.annotations.JavaTaskInfo;
import com.systemincloud.modeler.tasks.javatask.api.annotations.OutputPortInfo;
import com.systemincloud.modeler.tasks.javatask.api.data.Int32;
import com.systemincloud.modeler.tasks.javatask.api.data.Text;

@JavaTaskInfo(externalSource = true)
public class ExtSrcAtomicWith1SIn extends JavaTask {

	{
	    new Thread(new Runnable() {
			@Override
			public void run() {
				while(true) {
					try { Thread.sleep(2000);
					} catch (InterruptedException e) { }
					externalData();
				}
			}
		}).start();
	}
	
	@InputPortInfo(name = "In", dataType = Int32.class)
	public InputPort in;
	@OutputPortInfo(name = "Out", dataType = Text.class)
	public OutputPort out;

	@Override
	public void executeExtSrc() {
		out.putData(new Text("External data"));
	}

	@Override
	public void execute() {
		out.putData(new Text("From In: " + in.getData(Int32.class).getValue()));
	}

}

File: com.systemincloud.examples.tasks.pythontask/src/test/py/tasks/extsrc/ExtSrcAtomicWith1SIn.py
from sicpythontask.PythonTaskInfo import PythonTaskInfo
from sicpythontask.PythonTask import PythonTask
from sicpythontask.InputPort import InputPort
from sicpythontask.OutputPort import OutputPort
from sicpythontask.data.Int32 import Int32
from sicpythontask.data.Text import Text

import threading
import random

@PythonTaskInfo(external_source=True)
class ExtSrcAtomicWith1SIn(PythonTask):

    def __init_ports__(self):
        self.in_ = InputPort(name="in", data_type=Int32)
        self.out = OutputPort(name="out", data_type=[Text, Int32])

    def runner_start(self):
        self.notify()

    def notify(self):
        self.external_data()
        threading.Timer(1, self.notify).start()

    def execute_ext_src(self):
        self.out.put_data(Text('External: ' + str(random.getrandbits(100))))

    def execute(self, grp):
        inData = self.in_.get_data(Int32)
        self.out.put_data(Int32(10*inData.values[0]))

